// @launchdarkly/js-server-sdk-common/dist/events/EventFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const js_sdk_common_1 = require("@launchdarkly/js-sdk-common");
const isExperiment_1 = require("./isExperiment");
/**
 * @internal
 */
class EventFactory {
    constructor(withReasons) {
        this.withReasons = withReasons;
    }
    evalEvent(flag, context, detail, defaultVal, prereqOfFlag) {
        var _a;
        const addExperimentData = (0, isExperiment_1.default)(flag, detail.reason);
        return new js_sdk_common_1.internal.InputEvalEvent(this.withReasons, context, flag.key, defaultVal, detail, flag.version, 
        // Exclude null as a possibility.
        (_a = detail.variationIndex) !== null && _a !== void 0 ? _a : undefined, flag.trackEvents || addExperimentData, prereqOfFlag === null || prereqOfFlag === void 0 ? void 0 : prereqOfFlag.key, this.withReasons || addExperimentData ? detail.reason : undefined, flag.debugEventsUntilDate);
    }
    unknownFlagEvent(key, context, detail) {
        return new js_sdk_common_1.internal.InputEvalEvent(this.withReasons, context, key, detail.value, detail);
    }
    /* eslint-disable-next-line class-methods-use-this */
    identifyEvent(context) {
        return new js_sdk_common_1.internal.InputIdentifyEvent(context);
    }
    /* eslint-disable-next-line class-methods-use-this */
    customEvent(key, context, data, metricValue) {
        return new js_sdk_common_1.internal.InputCustomEvent(context, key, data !== null && data !== void 0 ? data : undefined, metricValue !== null && metricValue !== void 0 ? metricValue : undefined);
    }
}
exports.default = EventFactory;
