// @aws-sdk/client-elastic-load-balancing-v2/dist-cjs/ElasticLoadBalancingV2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticLoadBalancingV2 = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const AddListenerCertificatesCommand_1 = require("./commands/AddListenerCertificatesCommand");
const AddTagsCommand_1 = require("./commands/AddTagsCommand");
const CreateListenerCommand_1 = require("./commands/CreateListenerCommand");
const CreateLoadBalancerCommand_1 = require("./commands/CreateLoadBalancerCommand");
const CreateRuleCommand_1 = require("./commands/CreateRuleCommand");
const CreateTargetGroupCommand_1 = require("./commands/CreateTargetGroupCommand");
const DeleteListenerCommand_1 = require("./commands/DeleteListenerCommand");
const DeleteLoadBalancerCommand_1 = require("./commands/DeleteLoadBalancerCommand");
const DeleteRuleCommand_1 = require("./commands/DeleteRuleCommand");
const DeleteTargetGroupCommand_1 = require("./commands/DeleteTargetGroupCommand");
const DeregisterTargetsCommand_1 = require("./commands/DeregisterTargetsCommand");
const DescribeAccountLimitsCommand_1 = require("./commands/DescribeAccountLimitsCommand");
const DescribeListenerCertificatesCommand_1 = require("./commands/DescribeListenerCertificatesCommand");
const DescribeListenersCommand_1 = require("./commands/DescribeListenersCommand");
const DescribeLoadBalancerAttributesCommand_1 = require("./commands/DescribeLoadBalancerAttributesCommand");
const DescribeLoadBalancersCommand_1 = require("./commands/DescribeLoadBalancersCommand");
const DescribeRulesCommand_1 = require("./commands/DescribeRulesCommand");
const DescribeSSLPoliciesCommand_1 = require("./commands/DescribeSSLPoliciesCommand");
const DescribeTagsCommand_1 = require("./commands/DescribeTagsCommand");
const DescribeTargetGroupAttributesCommand_1 = require("./commands/DescribeTargetGroupAttributesCommand");
const DescribeTargetGroupsCommand_1 = require("./commands/DescribeTargetGroupsCommand");
const DescribeTargetHealthCommand_1 = require("./commands/DescribeTargetHealthCommand");
const ModifyListenerCommand_1 = require("./commands/ModifyListenerCommand");
const ModifyLoadBalancerAttributesCommand_1 = require("./commands/ModifyLoadBalancerAttributesCommand");
const ModifyRuleCommand_1 = require("./commands/ModifyRuleCommand");
const ModifyTargetGroupAttributesCommand_1 = require("./commands/ModifyTargetGroupAttributesCommand");
const ModifyTargetGroupCommand_1 = require("./commands/ModifyTargetGroupCommand");
const RegisterTargetsCommand_1 = require("./commands/RegisterTargetsCommand");
const RemoveListenerCertificatesCommand_1 = require("./commands/RemoveListenerCertificatesCommand");
const RemoveTagsCommand_1 = require("./commands/RemoveTagsCommand");
const SetIpAddressTypeCommand_1 = require("./commands/SetIpAddressTypeCommand");
const SetRulePrioritiesCommand_1 = require("./commands/SetRulePrioritiesCommand");
const SetSecurityGroupsCommand_1 = require("./commands/SetSecurityGroupsCommand");
const SetSubnetsCommand_1 = require("./commands/SetSubnetsCommand");
const ElasticLoadBalancingV2Client_1 = require("./ElasticLoadBalancingV2Client");
const commands = {
    AddListenerCertificatesCommand: AddListenerCertificatesCommand_1.AddListenerCertificatesCommand,
    AddTagsCommand: AddTagsCommand_1.AddTagsCommand,
    CreateListenerCommand: CreateListenerCommand_1.CreateListenerCommand,
    CreateLoadBalancerCommand: CreateLoadBalancerCommand_1.CreateLoadBalancerCommand,
    CreateRuleCommand: CreateRuleCommand_1.CreateRuleCommand,
    CreateTargetGroupCommand: CreateTargetGroupCommand_1.CreateTargetGroupCommand,
    DeleteListenerCommand: DeleteListenerCommand_1.DeleteListenerCommand,
    DeleteLoadBalancerCommand: DeleteLoadBalancerCommand_1.DeleteLoadBalancerCommand,
    DeleteRuleCommand: DeleteRuleCommand_1.DeleteRuleCommand,
    DeleteTargetGroupCommand: DeleteTargetGroupCommand_1.DeleteTargetGroupCommand,
    DeregisterTargetsCommand: DeregisterTargetsCommand_1.DeregisterTargetsCommand,
    DescribeAccountLimitsCommand: DescribeAccountLimitsCommand_1.DescribeAccountLimitsCommand,
    DescribeListenerCertificatesCommand: DescribeListenerCertificatesCommand_1.DescribeListenerCertificatesCommand,
    DescribeListenersCommand: DescribeListenersCommand_1.DescribeListenersCommand,
    DescribeLoadBalancerAttributesCommand: DescribeLoadBalancerAttributesCommand_1.DescribeLoadBalancerAttributesCommand,
    DescribeLoadBalancersCommand: DescribeLoadBalancersCommand_1.DescribeLoadBalancersCommand,
    DescribeRulesCommand: DescribeRulesCommand_1.DescribeRulesCommand,
    DescribeSSLPoliciesCommand: DescribeSSLPoliciesCommand_1.DescribeSSLPoliciesCommand,
    DescribeTagsCommand: DescribeTagsCommand_1.DescribeTagsCommand,
    DescribeTargetGroupAttributesCommand: DescribeTargetGroupAttributesCommand_1.DescribeTargetGroupAttributesCommand,
    DescribeTargetGroupsCommand: DescribeTargetGroupsCommand_1.DescribeTargetGroupsCommand,
    DescribeTargetHealthCommand: DescribeTargetHealthCommand_1.DescribeTargetHealthCommand,
    ModifyListenerCommand: ModifyListenerCommand_1.ModifyListenerCommand,
    ModifyLoadBalancerAttributesCommand: ModifyLoadBalancerAttributesCommand_1.ModifyLoadBalancerAttributesCommand,
    ModifyRuleCommand: ModifyRuleCommand_1.ModifyRuleCommand,
    ModifyTargetGroupCommand: ModifyTargetGroupCommand_1.ModifyTargetGroupCommand,
    ModifyTargetGroupAttributesCommand: ModifyTargetGroupAttributesCommand_1.ModifyTargetGroupAttributesCommand,
    RegisterTargetsCommand: RegisterTargetsCommand_1.RegisterTargetsCommand,
    RemoveListenerCertificatesCommand: RemoveListenerCertificatesCommand_1.RemoveListenerCertificatesCommand,
    RemoveTagsCommand: RemoveTagsCommand_1.RemoveTagsCommand,
    SetIpAddressTypeCommand: SetIpAddressTypeCommand_1.SetIpAddressTypeCommand,
    SetRulePrioritiesCommand: SetRulePrioritiesCommand_1.SetRulePrioritiesCommand,
    SetSecurityGroupsCommand: SetSecurityGroupsCommand_1.SetSecurityGroupsCommand,
    SetSubnetsCommand: SetSubnetsCommand_1.SetSubnetsCommand,
};
class ElasticLoadBalancingV2 extends ElasticLoadBalancingV2Client_1.ElasticLoadBalancingV2Client {
}
exports.ElasticLoadBalancingV2 = ElasticLoadBalancingV2;
(0, smithy_client_1.createAggregatedClient)(commands, ElasticLoadBalancingV2);
