// @launchdarkly/js-sdk-common/dist/internal/events/InputEvalEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class InputEvalEvent {
    constructor(withReasons, context, key, defValue, // default is a reserved keyword in this context.
    detail, version, variation, trackEvents, prereqOf, reason, debugEventsUntilDate) {
        var _a;
        this.key = key;
        this.kind = 'feature';
        this.creationDate = Date.now();
        this.context = context;
        this.default = defValue;
        this.variation = (_a = detail.variationIndex) !== null && _a !== void 0 ? _a : undefined;
        this.value = detail.value;
        if (version !== undefined) {
            this.version = version;
        }
        if (variation !== undefined) {
            this.variation = variation;
        }
        if (trackEvents !== undefined) {
            this.trackEvents = trackEvents;
        }
        if (prereqOf !== undefined) {
            this.prereqOf = prereqOf;
        }
        if (reason !== undefined) {
            this.reason = reason;
        }
        if (debugEventsUntilDate !== undefined) {
            this.debugEventsUntilDate = debugEventsUntilDate;
        }
    }
}
exports.default = InputEvalEvent;
